File: Python/Inflearn/4_자료구조 활용/review.py
# 자료구조 활용 (스택, 큐, 해시, 힙) review
# 🔥 가장 큰 수(스택)
import heapq as hq
from collections import deque


def solution(n, m):
    num = list(map(int, str(n)))
    stack = []

    for x in num:
        while stack and m > 0 and stack[-1] < x:
            stack.pop()
            m -= 1
        stack.append(x)

    if m != 0:
        stack = stack[:-m]

    return ''.join(map(str, stack))


if __name__ == "__main__":
    n, m = map(int, input().split())
    print(solution(n, m))


# 쇠막대기
def solution(exp):
    answer = 0
    stack = []
    for i in range(len(exp)):
        if exp[i] == '(':
            stack.append(exp[i])
        else:
            stack.pop()
            if exp[i-1] == ")":
                answer += 1
            else:
                answer += len(stack)
    return answer


if __name__ == "__main__":
    exp = input()
    print(solution(exp))


# 🔥 후위표기식 만들기: infix -> postfix (스택)
def solution(exp):
    answer = ''
    stack = []

    for x in exp:
        if x.isdecimal():
            answer += x
        else:
            if x == '(':
                stack.append(x)
            elif x == '*' or x == '/':
                while stack and (stack[-1] == '*' or stack[-1] == '/'):
                    answer += stack.pop()
                stack.append(x)
            elif x == '+' or x == '-':
                while stack and stack[-1] != '(':
                    answer += stack.pop()
                stack.append(x)
            elif x == ')':
                while stack and stack[-1] != '(':
                    answer += stack.pop()
                stack.pop()

    while stack:
        answer += stack.pop()

    return answer


if __name__ == "__main__":
    exp = input()
    print(solution(exp))


# 후위식 연산(스택)
def solution(exp):
    stack = []
    for x in exp:
        if x.isdecimal():
            stack.append(int(x))
        else:
            if x == '+':
                a = stack.pop()
                b = stack.pop()
                stack.append(b+a)
            elif x == '-':
                a = stack.pop()
                b = stack.pop()
                stack.append(b-a)
            elif x == '*':
                a = stack.pop()
                b = stack.pop()
                stack.append(b*a)
            elif x == '/':
                a = stack.pop()
                b = stack.pop()
                stack.append(b/a)
    return stack[0]


if __name__ == "__main__":
    exp = input()
    print(solution(exp))


# 공주 구하기(큐)
def solution(n, k):
    dq = deque(range(1, n+1))
    while len(dq) > 1:
        for _ in range(k-1):
            dq.append(dq.popleft())
        dq.popleft()
    return dq[0]


if __name__ == "__main__":
    n, k = map(int, input().split())
    print(solution(n, k))


# 응급실(큐)
def solution(n, m, danger):
    dq = deque([(idx, val) for idx, val in enumerate(danger)])
    cnt = 0
    while dq:
        temp = dq.popleft()
        if any(temp[1] < x[1] for x in dq):
            dq.append(temp)
        else:
            cnt += 1
            if temp[0] == m:
                return cnt


if __name__ == "__main__":
    n, m = map(int, input().split())
    danger = list(map(int, input().split()))
    print(solution(n, m, danger))


# 교육과정 설계(큐)
def solution(must, n, plan):
    dq = deque(must)
    for x in plan:
        if x in dq:
            if x != dq.popleft():
                return "NO"
    else:
        return "NO" if dq else "YES"


if __name__ == "__main__":
    must = input()
    n = int(input())
    for i in range(n):
        answer = solution(must, n, input())
        print(f'#{i+1} {answer}')


# 단어 찾기(해시)
def solution(n, words, poem):
    check = dict()
    for x in words:
        check[x] = check.get(x, 0) + 1
    for x in poem:
        check[x] -= 1
    # 다른 방법
    # for x in words:
    #   check[x] = 1
    # for x in poem:
    #   check[x] = 0
    for key, val in check.items():
        if val == 1:
            return key


if __name__ == "__main__":
    n = int(input())
    words = [input() for _ in range(n)]
    poem = [input() for _ in range(n-1)]
    print(solution(n, words, poem))


# 아나그램(딕셔너리 해시)
def solution(a, b):
    cnt = dict()
    for x in a:
        cnt[x] = cnt.get(x, 0) + 1
    for x in b:
        cnt[x] = cnt.get(x, 0) - 1
    for key, val in cnt.items():
        if val != 0:
            return "NO"
    else:
        return "YES"


if __name__ == "__main__":
    a = input()
    b = input()
    print(solution(a, b))


# 🔥 아나그램(리스트 해시)
def count(str):
    check = [0]*52
    for x in str:
        if x.isupper():
            check[ord(x)-65] += 1
        else:
            check[ord(x)-71] += 1
    return check


def solution(a, b):
    chA, chB = count(a), count(b)
    for i in range(52):
        if chA[i] != chB[i]:
            return "NO"
    else:
        return "YES"


if __name__ == "__main__":
    a = input()
    b = input()
    print(solution(a, b))


# 최소힙
if __name__ == "__main__":
    a = []
    while True:
        n = int(input())
        if n == -1:
            break
        if n == 0:
            if not a:
                print(-1)
            else:
                print(hq.heappop(a))
        else:
            hq.heappush(a, n)

# 최대힙
if __name__ == "__main__":
    a = []
    while True:
        n = int(input())
        if n == -1:
            break
        if n == 0:
            if not a:
                print(-1)
            else:
                print(-hq.heappop(a))
        else:
            hq.heappush(a, -n)

File: JavaScript/Inflearn/6_자료구조_스택, 큐/review.js
// 1. 올바른 괄호
const solution = (str) => {
  const stack = [];
  for (let item of str) {
    if (item === '(') stack.push();
    else if (item === ')') {
      if (stack.length === 0) return 'NO';
      stack.pop();
    }
  }
  if (stack.length === 0) return 'YES';
};

console.log(solution('(()(()))(()')); // NO

// 2. 괄호문자제거
const solution = (str) => {
  const stack = [];
  let answer = '';
  // 풀이1: answer에 문자 저장
  for (let item of str) {
    if (item === '(') stack.push('(');
    else if (item === ')') stack.pop();
    else {
      if (stack.length === 0) answer += item;
    }
  }

  // 풀이2: stack에 문자 저장
  for (item of str) {
    if (item === ')') {
      while (stack.pop() !== '(');
    } else stack.push(item);
  }
  answer = stack.join('');

  return answer;
};

console.log(solution('(A(BC)D)EF(G(H)(IJ)K)LM(N)')); // EFLM

// 3. 크레인 인형뽑기
const solution = (board, moves) => {
  let answer = 0;
  const stack = [];
  const boardLen = board.length;
  let num;
  for (let move of moves) {
    for (let i = 0; i < boardLen; i++) {
      if (board[i][move - 1]) {
        num = board[i][move - 1];
        board[i][move - 1] = 0;
        if (stack.length > 0 && stack[stack.length - 1] === num) {
          answer += 2;
          stack.pop();
        } else stack.push(num);
        break; // ✅ 인형 꺼냈으면 다음 moves로
      }
    }
  }

  return answer;
};

console.log(
  solution(
    [
      [0, 0, 0, 0, 0],
      [0, 0, 1, 0, 3],
      [0, 2, 5, 0, 1],
      [4, 2, 4, 4, 2],
      [3, 5, 1, 3, 1],
    ],
    [1, 5, 3, 5, 1, 2, 1, 4]
  )
); // 4

// 4. 후위식 연산(postfix)
const solution = (exp) => {
  const operate = (a, b, op) => {
    switch (op) {
      case '+':
        return a + b;
      case '-':
        return a - b;
      case '*':
        return a * b;
      case '/':
        return a / b;
    }
  };

  const nums = [];
  for (let item of exp) {
    if (isNaN(item)) {
      const num2 = nums.pop();
      const num1 = nums.pop();
      nums.push(operate(num1, num2, item));
    } else {
      nums.push(+item);
    }
  }

  return nums[0];
};

console.log(solution('352+*9-')); // 12

// 5. 쇠막대기
const solution = (exp) => {
  let answer = 0;
  const stack = [];
  for (let i = 0; i < exp.length; i++) {
    if (exp[i] === '(') stack.push(exp[i]);
    else {
      stack.pop();
      if (exp[i - 1] === '(') answer += stack.length;
      else answer++; // ✅ 쇠막대기 끝 의미
    }
  }

  return answer;
};

console.log(solution('()(((()())(())()))(())')); // 17
console.log(solution('(((()(()()))(())()))(()())')); // 24

// 6. 공주 구하기
const solution = (n, k) => {
  const queue = Array.from({ length: n }, (v, i) => i + 1);
  // 풀이1: count 변수로 k 카운트
  let count = 0;
  let current;
  while (queue.length > 1) {
    current = queue.shift();
    count++;
    if (count === k) count = 0;
    else queue.push(current);
  }
  return queue[0];

  // ✅ 풀이2: for문으로 k 카운트
  while (queue.length) {
    for (let i = 1; i < k; i++) queue.push(queue.shift());
    queue.shift();
    if (queue.length === 1) answer = queue.shift();
  }
  return answer;
};

console.log(solution(8, 3)); // 7

// 7. 교육과정 설계
const solution = (must, plan) => {
  const orders = must.split('');
  for (let subject of plan) {
    // 풀이1: 첫 번째 원소와 일치하면 shift (끝까지 순회하므로 비효율적))
    if (subject === orders[0]) {
      orders.shift();
    }
    // 풀이2: includes()로 큐에 있는지 확인한 후 shift
    // ✅ shift한 값이 subject와 일치하지 않으면 바로 종료
    if (orders.includes(subject)) {
      if (subject !== orders.shift()) return 'NO';
    }
  }
  return orders.length ? 'NO' : 'YES';
};

console.log(solution('CBA', 'CBDAGE')); // YES
